consensus/core/src/leader_scoring_strategy.rs

```diff
@@ -0,0 +1,457 @@
+/// the vote.
+pub(crate) struct CertifiedVoteScoringStrategyV2 {}
+
+impl ScoringStrategy for CertifiedVoteScoringStrategyV2 {
+    fn calculate_scores_for_leader(
+        &self,
+        subdag: &UnscoredSubdag,
+        leader_slot: Slot,
+        committer: &BaseCommitter,
+    ) -> Vec<u64> {
+        let num_authorities = subdag.context.committee.size();
+        let mut scores_per_authority = vec![0_u64; num_authorities];
+
+        let wave = committer.wave_number(leader_slot.round);
+        let decision_round = committer.decision_round(wave);
+
+        let leader_blocks = subdag.get_blocks_at_slot(leader_slot);
+
+        if leader_blocks.is_empty() {
+            tracing::info!("[{}] No block for leader slot {leader_slot} in this set of unscored committed subdags, skip scoring", subdag.context.own_index);
+            return scores_per_authority;
+        }
+
+        // At this point we are guaranteed that there is only one leader per slot
+        // because we are operating on committed subdags.
+        assert!(leader_blocks.len() == 1);
+
+        let leader_block = leader_blocks.first().unwrap();
+
+        let decision_blocks = subdag.get_blocks_at_round(decision_round);
+
+        let mut all_votes: HashMap<BlockRef, (bool, StakeAggregator<QuorumThreshold>)> =
+            HashMap::new();
+        for potential_cert in decision_blocks {
+            let authority = potential_cert.reference().author;
+            for reference in potential_cert.ancestors() {
+                if let Some((is_vote, stake_agg)) = all_votes.get_mut(reference) {
+                    if *is_vote {
+                        stake_agg.add(authority, &subdag.context.committee);
+                    }
+                } else if let Some(potential_vote) = subdag.get_block(reference) {
+                    let is_vote = subdag.is_vote(&potential_vote, leader_block);
+                    let mut stake_agg = StakeAggregator::<QuorumThreshold>::new();
+                    stake_agg.add(authority, &subdag.context.committee);
+                    all_votes.insert(*reference, (is_vote, stake_agg));
+                } else {
+                    tracing::info!(
+                        "Potential vote not found in unscored committed subdags: {:?}",
+                        reference
+                    );
+                };
+            }
+        }
+
+        for (vote_ref, (is_vote, stake_agg)) in all_votes {
+            if is_vote {
+                let authority = vote_ref.author;
+                tracing::info!(
+                    "Found a certified vote {vote_ref} for leader {leader_block} from authority {authority}"
+                );
+                tracing::info!(
+                    "[{}] scores +{} reputation for {authority}!",
+                    subdag.context.own_index,
+                    stake_agg.stake()
+                );
+                scores_per_authority[authority] += stake_agg.stake();
```

Comment:
This will bias the scores to validators already with higher stake. Let's save this strategy else where and iterate on it more before checking it in.